dbms/src/DataStreams/copyData.cpp

```diff
@@ -51,6 +54,74 @@ void copyDataImpl(IBlockInputStream & from, IBlockOutputStream & to, TCancelCall
 
 inline void doNothing(const Block &) {}
 
+
+void copyDataImpl(BlockInputStreams & froms, BlockOutputStreams & tos)
+{
+    if (froms.size() == tos.size())
+    {
+        std::vector<ThreadFromGlobalPool> threads;
+        threads.reserve(froms.size());
+        for (size_t i = 0; i < froms.size(); i++)
+        {
+            threads.emplace_back([from = froms.at(i), to = tos.at(i)]()
+            {
+                from->readPrefix();
+                to->writePrefix();
+                while (Block block = from->read())
+                    to->write(block);
+                from->readSuffix();
+                to->writeSuffix();
+            });
+        }
+        for (auto & thread : threads)
+            thread.join();
+    }
+    else
+    {
+        ConcurrentBoundedQueue<Block> queue(froms.size());
+        ThreadFromGlobalPool from_threads([&]()
```

Comment:
I think we can avoid this extra thread.
But now it's also ok.

Reply:
This extra thread is void in new commit by swap the creation order of reader and writer.